File: C-RandChar/src/RastgeleKarakter.c
/**
*
* @author Omer Kirsever omer.kirsever@ogr.sakarya.edu.tr 
* @since 21.04.2019
* Rastgele karakter ve kelimeler uretebilen kutuphanenin kodlari.
* 
*/
#include "RastgeleKarakter.h"

// RastgeleKarakter sinifi icin olusturma fonksiyonu
RastgeleKarakter RastgeleKarakterSinifOlustur(){
	RastgeleKarakter this;
	this = (RastgeleKarakter)malloc(sizeof(struct RASTGELEKARAKTER));
	this->rand = RandomSinifKur();
	return this;
}
// Random.c de hazirlanan islemler sonunda rastgele sayi uretip chara ceviren fonksiyon
char RandomKarakter(RastgeleKarakter rk){
	unsigned sayi = RandomSayi(rk->rand)%122; 
	while(sayi < 65 || (sayi < 97 && sayi > 90))
		sayi = RandomSayi(rk->rand)%122;
	return (char)sayi;	
}

// Sayili rastgele karakter basan fonksiyon. 
char* SayiliRandomKarakter(RastgeleKarakter rk,int miktar)
{   
	char *dizi = (char*)malloc(sizeof(char)*miktar);
	for(int i = 0; i < miktar; i++){
		dizi[i] = RandomKarakter(rk);
	}
	return dizi;
}

//Verilen aralik arasinda tek karakter basan fonksiyon.
char AralikliRastgele(RastgeleKarakter rk, char ilkharf,char sonharf){
int ilkharf2=ilkharf+1;
int aralik= sonharf-ilkharf2;
unsigned sayi = RandomSayi(rk->rand)%aralik;
int yenisayi=sayi+ilkharf2;
return (char)yenisayi;
}

//Verilen aralik arasinda istenilen miktarda karakter basan fonksiyon.
char* SayiliAralikliRastgele(RastgeleKarakter rk, char ilkharf,char sonharf,int miktar){
	int ilkharf2=ilkharf+1;
	int aralik= sonharf-ilkharf2;
	char *dizi = (char*)malloc(sizeof(char)*miktar);
	for(int i = 0; i < miktar; i++){
		unsigned sayi = RandomSayi(rk->rand)%aralik;
		int yenisayi=sayi+ilkharf2;
		dizi[i]=yenisayi;
	}
	dizi[miktar]='\0';
	return dizi;
}
//Belirtilen charlar arasinda rastgele istenilen miktarda karakter basan fonksiyon.
char* BelirtilenKarakterler(RastgeleKarakter rk,int miktar,char asci1,char asci2,char asci3, char asci4,char asci5,char asci6){
	char *dizi = (char*)malloc(sizeof(char)*miktar);
	for(int i = 0; i < miktar; i++){
		unsigned sayi = RandomSayi(rk->rand)%6;
		if(sayi==0){
			dizi[i]=asci1;
		}	
			else if(sayi==1){
			dizi[i]=asci2;
		}	
			else if(sayi==2){
			dizi[i]=asci3;
		}	
			else if(sayi==3){
			dizi[i]=asci4;
		}	
			else if(sayi==4){
			dizi[i]=asci5;
		}
			else if(sayi==5){
			dizi[i]=asci6;
		}		
	}
	dizi[miktar]='\0';
	return dizi;
}

//Kelime sayisi ve kelimedeki min-max harf sayisini fonksiyonda alip rastgele cumle ureten fonksiyon. 
char* RastgeleCumle(RastgeleKarakter rk,int kelimesayisi,int maxkelimeharf,int minkelimeharf){
	int x=0;
	int gecmistoplam=0;
	char *dizi = (char*)malloc(sizeof(char)*x);
	
	for (int i = 0; i < kelimesayisi; i++){
		unsigned yenisayi = RandomSayi(rk->rand)%maxkelimeharf;
		if(yenisayi<minkelimeharf){
			yenisayi=minkelimeharf;
		}
		gecmistoplam=gecmistoplam+yenisayi;
		dizi[x]=' ';
			for (int x=(gecmistoplam-yenisayi+1); x <gecmistoplam; x++){
				dizi[x] = RandomKarakter(rk);
			}
				x=gecmistoplam;
	}
	dizi[x]='\0';
	return dizi;
}

// Sinifi yokeden fonksiyon
void RastgeleKarakterSinifYoket(RastgeleKarakter rk){
	if(rk == NULL) return;
	free(rk->rand);
	rk->rand = NULL;
	free(rk);
	rk = NULL;
}

File: java-RandChar/Rastgele/src/paket/RastgeleKarakter.java
/**
*
* @author Omer Kirsever omer.kirsever@ogr.sakarya.edu.tr , Batuhan Ucan batuhan.ucan@ogr.sakarya.edu.tr
* @since 16.03.2019
* Rastgele karakter ve kelimeler uretebilen kutuphanenin test kodlari.
* 
*/
package paket;


public class RastgeleKarakter {
    int sayi;
    char harf;
    int yenisayi=0;
    public char RastgeleHarf(){  //tek karakter basmasi icin
        
        sayi = (int)System.nanoTime(); // System saatini int sayi degiskenine atiyoruz.
        if(sayi<0) // System saati bazen negatif cikiyor. Onu pozitife ceviriyoruz.
        {             
            sayi=sayi*(-1);   
        } 
        // Buyuk kucuk harf rastgele bassin diye ilk mod 2 aliyoruz. Ordan gelen sayi 0 ise buyuk harf 1 ise kucuk harf. 
        if(sayi%2==0)  
        {  
            sayi=(sayi% 26)+65;      // Burada buyuk harf basiliyor.
            harf = (char)sayi;      
        }
        else
        {
            sayi=(sayi% 26)+97;      //Burada ise kucuk harf basiliyor.
             
        } 
         harf = (char)sayi;  
         return harf;
    } 

    public String SayiliRastgele(int adet){ // fonksiyondaki adet miktari kac tane sayi basilicak adeti
        String a = "";
        for(int x=0;x<adet;x++)
        {
            a += RastgeleHarf();// ustteki ana fonksiyon cagiriliyor.
        }
        return a ;
    }
    
    public String AralikliRastgele(int ilkharf, int sonharf,int adet){ 
        String a = "";
        char b;
       // Verilen ascii sayilari arasinda verilen adet kadar harf basma
      int ilkharf2 = ilkharf+1; // Verilen ilk harfin ve son harfin haricinde, icindeki harfleri yazmasi icin.
      int aralik=0;
      for(int x=0;x<adet;x++){ //kac tane harf basilacaksa o kadar for donduruyor.
      sayi = (int)System.nanoTime();
      if(sayi<0)
      {
          sayi=sayi*(-1);
      } 
      aralik=(sonharf-ilkharf2); // O iki harf arasinda kac tane harf oldugunu ogreniyoruz.
      sayi=(sayi % aralik)+(ilkharf2); // sayiyi araliga modunu alarak ilk girilen harf+1 ile toplayip yeni sayi uretiyoruz.
      harf=(char)sayi;          //sayiyi chara ceviriyoruz. 
      a += harf;
      //System.out.print(harf);    //harfi yazdiriyoruz.
      }
      return a;
    }
    
    public String BelirtilenKarakter(int adet,int a,int b,int c,int d,int e,int f){ 
        String z = "";
        // verilen adet ve girilen asciiler arasinda sayi yazdirma
      for(int x=0;x<adet;x++) // girilen adet kadar for döndürme
      {
        sayi = (int)System.nanoTime();// System saatini int sayi degiskenine atiyoruz.
        if(sayi<0)// System saati bazen negatif cikiyor. Onu pozitife ceviriyoruz.
        {
            sayi=sayi*(-1);
        } 
        sayi=(sayi%6); 
        // sayiyi girilen sayi adetinde mod alıp mod degerine karsılık gelen sirayi yazdiriyoruz.
            if(sayi==0)
            {
                sayi=a;
            }
            else if(sayi==1)
            {
                sayi=b;
            }
            else if(sayi==2)
            {
                sayi=c;
            }
            else if(sayi==3)
            {
                sayi=d;
            }
            else if(sayi==4)
            {
                sayi=e;
            } 
            else if(sayi==5)
            {
                sayi=f;
            }
        harf=(char)sayi;
        //System.out.print(harf);
        z += harf;
      }
      return z;
    }   
    public String CumleKur(int kelimesayisi,int minkelimeharf, int maxkelimeharf){
        String a = "";
      //Fonksiyonda cumledeki kelime sayisi, kelimedeki min harf ve max harf sayisi aliniyor.
        for(int x=0;x<kelimesayisi;x++) //yazilacak kelime sayisi kadar for dongusu
        {
            yenisayi = (int)System.nanoTime(); // aynı sistem saati islemleri 
                if(yenisayi<0)
                {
                    yenisayi=yenisayi*(-1);
                } 
            yenisayi=(yenisayi%maxkelimeharf); // sistem saati basilacak maxkelime uzunlugunla modu aliniyor.
                if(yenisayi<minkelimeharf) //eger sayi girilen min harften kucukse degeri min harfe eşitliyor. 
                {
                    yenisayi=minkelimeharf; // yani min harf sayisindan daha az harfli kelime uretmemek icin
                }
                    for(int y=0;y<yenisayi;y++) // daha sonra o sayi kadar harf basiyor bosluk bırakıyor üstteki fora giriyor.
                    {
                        a += RastgeleHarf(); //ustteki daha once hazirladigimiz ana fonksiyonu kullaniyoruz. 
                    }
                    a += " ";
        } 
         return a ;
    }     
    
}
